daemon/session_agent.go

```diff
@@ -0,0 +1,88 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+type SessionAgent struct {
+	Version  string
+	listener net.Listener
+	serve    *shutdownServer
```

Comment:
`shutdownServer` was written because we needed to work with go 1.6; since 1.8 go's regular `http.Server` has [`Shutdown`](https://golang.org/pkg/net/http/#Server.Shutdown) which AFAIK implements what we needed, we just haven't had time to move daemon over.
Unfortunately the semantics are a little bit different so it's not (again AFAIK) a simple refactor.

However, spreading the madness™ is not a reasonable way forwards.

Do you think you could use the regular http.Server's shutdown bits for what you need?

Reply:
I'll have a go.  From the look of it, I should be able to replace:

    s.tomb.Kill(s.serve.finishShutdown())

... with:

    context, _ := context.WithTimeout(s.tomb.Context(nil), shutdownTimeout)
    s.tomb.Kill(s.serve.Shutdown(context))

I think that should give the same semantics, and the new timeout context should be cancelled by the Kill call.